linalg/auxiliary.cpp

```diff
@@ -0,0 +1,512 @@
+   but we recommend the CG wrapper for robustness here. */
+MatrixFreeAuxiliarySpace::MatrixFreeAuxiliarySpace(
+   ParMesh& mesh_lor, Coefficient* beta_coeff,
+   MatrixCoefficient* beta_mcoeff, Array<int>& ess_bdr, Operator& curlcurl_oper,
+   Operator& g,
+#ifdef MFEM_USE_AMGX
+   bool useAmgX_,
+#endif
+   int cg_iterations)
+   :
+   Solver(curlcurl_oper.Height()),
+   comm(mesh_lor.GetComm()),
+   matfree(NULL),
+   cg(NULL),
+#ifdef MFEM_USE_AMGX
+   useAmgX(useAmgX_),
+#endif
+   inner_aux_iterations(0)
+{
+   H1_FECollection * fec_lor = new H1_FECollection(1, mesh_lor.Dimension());
+   ParFiniteElementSpace fespace_lor(&mesh_lor, fec_lor);
+
+   // build LOR AMG v-cycle
+   ParBilinearForm a_space(&fespace_lor);
+
+   // we need something like DIAG_ZERO in the solver, but explicitly doing
+   // that makes BoomerAMG setup complain, so instead we constrain the boundary
+   // in the CG solver
+   const Matrix::DiagonalPolicy policy = Matrix::DIAG_ONE;
+
+   a_space.SetDiagonalPolicy(policy);
+
+   if (beta_mcoeff != NULL)
+   {
+      MFEM_VERIFY(beta_coeff == NULL, "Only one beta coefficient should be defined.");
+      a_space.AddDomainIntegrator(new DiffusionIntegrator(*beta_mcoeff));
+   }
+   else if (beta_coeff != NULL)
+   {
+      a_space.AddDomainIntegrator(new DiffusionIntegrator(*beta_coeff));
+   }
+   else
+   {
+      a_space.AddDomainIntegrator(new DiffusionIntegrator);
+   }
+   a_space.UsePrecomputedSparsity();
+   a_space.Assemble();
+   if (ess_bdr.Size())
+   {
+      fespace_lor.GetEssentialTrueDofs(ess_bdr, ess_tdof_list);
+   }
+
+   // you have to use (serial) BilinearForm eliminate routines to get
+   // diag policy DIAG_ZERO all the ParallelEliminateTDofs etc. routines
+   // implicitly have a Matrix::DIAG_KEEP policy
+   a_space.EliminateEssentialBC(ess_bdr, policy);
+   a_space.Finalize();
+   aspacematrix = a_space.ParallelAssemble();
+
+   aspacematrix->CopyRowStarts();
+   aspacematrix->CopyColStarts();
+
+   SetupAMG(0);
+
+   if (cg_iterations > 0)
+   {
+      SetupCG(curlcurl_oper, g, cg_iterations);
+   }
+   else
+   {
+      SetupVCycle();
+   }
+
+   delete fec_lor;
+}
+
+void MatrixFreeAuxiliarySpace::SetupCG(
+   Operator& curlcurl_oper, Operator& conn,
+   int inner_cg_iterations)
+{
+   MFEM_ASSERT(conn.Height() == curlcurl_oper.Width(),
+               "Operators don't match!");
+   matfree = new RAPOperator(conn, curlcurl_oper, conn);
+   MFEM_ASSERT(matfree->Height() == aspacepc->Height(),
+               "Operators don't match!");
+
+   cg = new CGSolver(comm);
+   cg->SetOperator(*matfree);
+   cg->SetPreconditioner(*aspacepc);
+   if (inner_cg_iterations > 99)
+   {
+      cg->SetRelTol(1.e-14);
+      cg->SetMaxIter(100);
+   }
+   else
+   {
+      cg->SetRelTol(0.0);
+      cg->SetMaxIter(inner_cg_iterations);
+   }
+   cg->SetPrintLevel(-1);
+
+   aspacewrapper = cg;
+}
+
+void MatrixFreeAuxiliarySpace::SetupVCycle()
+{
+   aspacewrapper = aspacepc;
+}
+
+class ZeroWrap : public Solver
```

Comment:
It would be more accurate to name this something like `ZeroWrapAMG`.

Reply:
Done.